test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,316 @@
+	pb.WithArgs([]string{"-conf", "/etc/coredns/Corefile"})
+	pb.AddVolume(volume)
+	pb.AddVolumeMount(volumeMount)
+
+	require.NoError(t, pb.Create(data))
+	require.NoError(t, data.podWaitForRunning(defaultTimeout, customDnsPodName, data.testNamespace))
+	require.NoError(t, data.patchPodAnnotation(data.testNamespace, customDnsPodName, nil), "failed to annotate the custom dns pod.")
+
+}
+
+func buildFqdnPolicy(t *testing.T, data *TestData) *utils.AntreaNetworkPolicySpecBuilder {
+	podSelectorLabel := map[string]string{
+		fqdnPodSelectorLabelKey: fqdnPodSelectorLabelValue,
+	}
+	port := int32(80)
+	udpPort := int32(53)
+	builder := &utils.AntreaNetworkPolicySpecBuilder{}
+	builder = builder.SetName(data.testNamespace, fqdnPolicyName).
+		SetTier(defaultTierName).
+		SetPriority(1.0).
+		SetAppliedToGroup([]utils.ANNPAppliedToSpec{{PodSelector: podSelectorLabel}})
+	builder.AddFQDNRule(testFullyQualifiedDomainName, utils.ProtocolTCP, &port, nil, nil, "AllowForFQDN", nil,
+		crdv1beta1.RuleActionAllow)
+	builder.AddEgress(utils.ProtocolUDP, &udpPort, nil, nil, nil, nil,
+		nil, nil, nil, nil, map[string]string{customDnsLabelKey: customDnsLabelValue},
+		nil, nil, nil, nil,
+		nil, nil, crdv1beta1.RuleActionAllow, "", "AllowDnsQueries")
+	builder.AddEgress(utils.ProtocolTCP, nil, nil, nil, nil, nil,
+		nil, nil, nil, nil, nil,
+		nil, nil, nil, nil,
+		nil, nil, crdv1beta1.RuleActionReject, "", "DropAllRemainingTraffic")
+
+	annp, err := k8sUtils.CreateOrUpdateANNP(builder.Get())
+	require.NoError(t, err, "error while deploying antrea policy %+v", err)
+	failOnError(err, t)
+	failOnError(waitForResourceReady(t, 30*time.Second, annp), t)
+
+	return builder
+}
+
+func createToolBoxPod(t *testing.T, data *TestData, dnsServiceIP string) {
+	toolBoxLabel := map[string]string{fqdnPodSelectorLabelKey: fqdnPodSelectorLabelValue}
+	pb := NewPodBuilder(toolBoxPodName, data.testNamespace, ToolboxImage)
+	pb.WithLabels(toolBoxLabel)
+	pb.WithContainerName(toolboxContainerName)
+	mutateSpecForAddingCustomDNS := func(pod *corev1.Pod) {
+		pod.Spec.DNSPolicy = corev1.DNSNone
+		if pod.Spec.DNSConfig == nil {
+			pod.Spec.DNSConfig = &corev1.PodDNSConfig{}
+		}
+		pod.Spec.DNSConfig.Nameservers = []string{dnsServiceIP}
+
+	}
+	pb.WithMutateFunc(mutateSpecForAddingCustomDNS)
+	require.NoError(t, pb.Create(data))
+	require.NoError(t, data.podWaitForRunning(defaultTimeout, toolBoxPodName, data.testNamespace))
+}
+
+func createHttpAgnhostPod(t *testing.T, data *TestData, podName string, agnLabels map[string]string) *PodIPs {
+	args := []string{"netexec", "--http-port=" + strconv.Itoa(agnHostPort)}
+	ports := []corev1.ContainerPort{
+		{
+			Name:          "http",
+			ContainerPort: agnHostPort,
+			Protocol:      corev1.ProtocolTCP,
+		},
+	}
+
+	require.NoError(t, NewPodBuilder(podName, data.testNamespace, agnhostImage).WithArgs(args).WithPorts(ports).WithLabels(agnLabels).Create(data))
+	podIP, err := data.podWaitForIPs(defaultTimeout, podName, data.testNamespace)
```

Comment:
keep `podIPs` as the variable name, not `podIP` (same in the caller function), otherwise it can get confusing pretty quickly. There may be 2 IPs (IPv4 + IPv6) for dual stack clusters